`Programming exercises`: Combine reset actions into one button

FelixTJDietrich opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
#### General
<!-- You only need to choose one of the first two check items: Generally, test on the test servers. -->
<!-- If it's only a small change, testing it locally is acceptable and you may remove the first checkmark. If you are unsure, please test on the test servers. -->
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/guidelines/development-process/#naming-conventions-for-github-pull-requests).


#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I implemented the changes with a good performance and prevented too many database calls.
- [x] I documented the Java code using JavaDoc style.

#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/) and ensured that the layout is responsive.
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

#### Changes affecting Programming Exercises
- [x] I tested **all** changes and their related features with **all** corresponding user types on Test Server 1 (Atlassian Suite).
- [x] I tested **all** changes and their related features with **all** corresponding user types on Test Server 2 (Jenkins and Gitlab).

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->

@krusche noted that the `Build Plan Reset` and `Cleanup` on the exercise detail and the `Reset` button in the exercise table should be consolidated into one button that opens a dialog to handle all three operations at once. This fixes inconsistencies across the pages and removes the number of buttons for an overall better UX.

Current `Build Plan Reset` and `Cleanup` buttons on the exercise detail:
![image_720](https://user-images.githubusercontent.com/5898705/228945466-1b358ddb-24ea-43ae-9452-f3edf0e6df73.png)

Current `Reset` button on the exercise table:
<img width="420" alt="image" src="https://user-images.githubusercontent.com/5898705/228946620-0124a630-dddc-4552-983c-506ee79e4c2d.png">


### Description
<!-- Describe your changes in detail -->

Consolidated the three buttons into one `Reset` button and added it to the exercise detail and table pages. The button opens a dialog with the following options:
[ ] Delete student build plans [Instructors]
[ ] Delete student repositories [Instructors]
[ ] Delete all student participations, submissions and results [Instructors]
[ ] Recreate the exercise build plans (BASE and SOLUTION) again from the latest exercise template, overwriting any previous changes [Editors + Instructors]

An info box appears when selecting `Delete student build plans` that states that build plans will automatically recreate, this info box disappears if also `Delete student repositories` is enabled.
Selecting `Delete student repositories` also automatically selects `Delete student build plans` since deleting repositories without the build plans does not make much sense (I also wanted to reuse the `cleanup` function that was already provided that required this).
If option 3 is selected without all prior options enabled, a warning appears to warn about potential CI or VC artefacts.

Added CI (Bamboo/Jenkins) and VC (Bitbucket/Gitlab) names to the info endpoint and added their names to the dialog.

Added a PUT `reset` endpoint for resetting a programming exercise `/programming-exercises/{exerciseId}/reset` that receives a DTO with the dialog options and consolidates the following endpoints into one:
* `/programming-exercises/{exerciseId}/recreate-build-plans`
* `exercises/{exerciseId}/reset`
* `exercises/{exerciseId}/cleanup`

Deleted the now unused `/cleanup` and `/recreate-build-plans` endpoints.

### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->
Prerequisites:
- 1 Editor
- 1 Instructor
- 1 Student
- 1 Programming Exercise with a student participation

1. Log into Artemis as an Instructor
2. Navigate to Course Management
3. Click on a programming exercise to navigate to the exercise detail
4. Verify that `Build Plan Reset` and `Cleanup` buttons are removed and a new `Reset` button exists
5. Click on `Reset`, it should open a `Reset Programming Exercise` dialog
6. Click on `Cancel` and check if the dialog closes
7. Press `Reset` to reopen the dialog again
9. Play around with the options (Option 2 requires option 1 to be enabled, done automatically)
10. Verify that the text makes sense and the info box and warning box that appear also make sense
11. A instructor should only be able to press `Reset` if the correct exercise name is typed in the confirmation box and at least one option is picked
12. Verify that it shows the correct name of the current Artemis configuration's CI and VC
13. Check out the participations in the course (open in separate tab, `Participation` button on exercise detail)
14. Find a participation with a build plan id (press refresh to recreate one, or create a new submission with a student)
15. Click the build plan id and open it on the CI platform in a separate tab
16. Open the dialog and reset with option `Delete student build plans` selected
17. Refresh participations and the build plans should be gone now
18. The build plan should also be gone on CI
19. Open a repository url on the VC platform in a separate tab (also in the participations screen)
20. Open the dialog and reset with option `Delete student repositories` (requires that 1. is set, will set and unset automatically)
21. Refresh participations and the repository urls and build plan ids should be gone
22. Verify on the VC platform that the repository has been removed
23. Open the dialog and reset with option `Delete all student participations, submissions and results`
24. Refresh participations, they should all be gone now
25. On exercise details, scroll down and open the solution and template build plan on the CI platform in separate tabs
26. Open the dialog and reset with option `Recreate the exercise build plans (BASE and SOLUTION) again from the latest exercise template, overwriting any changes`
27. Refresh the solution and template build plan on the CI platform should just have been created

**❗️There are no editor accounts on the test servers AFAIK, I tested it locally and it works. You can skip this for now (?).**
28. Log into Artemis as an Editor
29. Perform a `Reset` operation and verify that only `Recreate the exercise build plans (BASE and SOLUTION) again from the latest exercise template, overwriting any changes` exists
30. Perform that operation and verify that is succeeded 

### Review Progress
<!-- Each Pull Request should be reviewed by at least two other developers. The code, the functionality (= manual test) and the exam mode need to be reviewed. -->
<!-- The reviewer or author check the following boxes depending on what was reviewed or tested. All boxes should be checked before merge. -->
<!-- You can add additional checkboxes if it makes sense to only review parts of the code or functionality. -->
<!-- When changes are pushed, uncheck the affected boxes. (Not all changes require full re-reviews.) -->
<!-- All PRs that might affect the exam mode (e.g. change a client component that is also used in the exam mode) need an additional verification that the exam mode still works. -->

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage
<!-- Please add the test coverages for all changed files here. You can see this when executing the tests locally (see build.gradle and package.json) or when looking into the corresponding Bamboo build plan. -->
<!-- Lines are the main reference but a significantly lower branch percentage can indicate missing edge cases in the tests. -->
<!-- Note: You may use the table below or copy the file coverage from the Codecov bot's comment. -->
<!--
| Class/File | Branch | Line |
|------------|-------:|-----:|
| ExerciseService.java | 85% | 77% |
| programming-exercise.component.ts | 13% | 95% |
-->

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [programming-exercise-detail.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4716-22/artifact/shared/Coverage-Report-Client-Tests/app/exercises/programming/manage/programming-exercise-detail.component.ts.html) | 67.13% | ✅ |
| [programming-exercise.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4716-22/artifact/shared/Coverage-Report-Client-Tests/app/exercises/programming/manage/programming-exercise.component.ts.html) | 90.29% | ✅ |
| [programming-exercise-reset-button.directive.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4716-22/artifact/shared/Coverage-Report-Client-Tests/app/exercises/programming/manage/reset/programming-exercise-reset-button.directive.ts.html) | N/A | (not much to test) ❌ |
| [programming-exercise-reset-dialog.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4716-22/artifact/shared/Coverage-Report-Client-Tests/app/exercises/programming/manage/reset/programming-exercise-reset-dialog.component.ts.html) | 77.5% | ✅ |
| [programming-exercise.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4716-22/artifact/shared/Coverage-Report-Client-Tests/app/exercises/programming/manage/services/programming-exercise.service.ts.html) | 61.6% | ✅ |

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [ExerciseResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4716-22/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest/ExerciseResource.html) | 95% | (deleted `/cleanup`) ✅ |
| [ProgrammingExerciseResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4716-22/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest/ProgrammingExerciseResource.html) | 90% |  ✅ |
| [ProgrammingExerciseResetOptionsDTO.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4716-22/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.dto/ProgrammingExerciseResetOptionsDTO.html) | 100% | ✅ |


### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->

#### Editor
Buttons for editor on the exercise detail (`Reset` replaces `Build Plan Reset`):
<img width="423" alt="image" src="https://user-images.githubusercontent.com/5898705/235961937-e5b95f93-f3a1-4b03-9c23-614d7afd25ea.png">

Buttons for editor on the exercise table (`Reset` has beed added):
<img width="1449" alt="image" src="https://user-images.githubusercontent.com/5898705/235964100-434e24bb-4948-479a-9a76-3f8296be2e7c.png">

Reset dialog for editors:
<img width="805" alt="image" src="https://user-images.githubusercontent.com/5898705/235961995-3cc9935d-9cff-475c-84cf-f3dc7253a5b4.png">

#### Instructor
Buttons for instructor on the exercise detail (`Reset` replaces `Build Plan Reset` and `Cleanup`):
<img width="595" alt="image" src="https://user-images.githubusercontent.com/5898705/232352699-2d681abe-ebe3-468c-906a-fa8b65dfc5e9.png">

Reset dialog with no option selected for instructors:
<img width="1095" alt="image" src="https://user-images.githubusercontent.com/5898705/232351846-c335c950-d525-44ca-abf9-e918b70d7a0f.png">

Reset dialog with infos and warnings when some options are set for instructors:
<img width="806" alt="Screenshot 2023-04-27 at 14 43 25" src="https://user-images.githubusercontent.com/5898705/234865650-753c411d-9b83-4ecc-8211-7ed10ec1e1ea.png">

Reset dialog in German and light mode for instructors: 
<img width="809" alt="Screenshot 2023-04-27 at 14 43 40" src="https://user-images.githubusercontent.com/5898705/234865673-ba5891e9-960e-4fd9-aa4b-7d2ed7ff113c.png">

JohannesStoehr commented on src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java:
```diff
@@ -625,6 +626,45 @@ public ResponseEntity<Void> recreateBuildPlans(@PathVariable Long exerciseId) {
+        log.debug("REST request to reset ProgrammingExercise : {}", exerciseId);
+        var programmingExercise = programmingExerciseRepository.findByIdWithTemplateAndSolutionParticipationAndAuxiliaryRepositoriesElseThrow(exerciseId);
+
+        User user = userRepository.getUserWithGroupsAndAuthorities();
+        authCheckService.checkHasAtLeastRoleForExerciseElseThrow(Role.INSTRUCTOR, programmingExercise, user);
```
```suggestion
        authCheckService.checkHasAtLeastRoleForExerciseElseThrow(Role.INSTRUCTOR, programmingExercise, null);
```

JohannesStoehr commented on src/main/java/de/tum/in/www1/artemis/web/rest/dto/ProgrammingExerciseResetOptionsDTO.java:
```diff
@@ -0,0 +1,50 @@
+ * This is a dto for the programming exercise reset options.
+ */
+@JsonInclude(JsonInclude.Include.NON_EMPTY)
+public class ProgrammingExerciseResetOptionsDTO {
```
How about using a record?

FelixTJDietrich replied:
TIL, will use record

JohannesStoehr commented on src/main/webapp/app/exercises/programming/manage/programming-exercise-detail.component.html:
```diff
@@ -182,9 +182,8 @@ <h2><span jhiTranslate="artemisApp.programmingExercise.detail.title">Programming
-
                     <button
-                        *ngIf="programmingExercise.isAtLeastEditor"
+                        *ngIf="programmingExercise.isAtLeastEditor && !programmingExercise.isAtLeastInstructor"
```
Why should instructors not have this option?
Is this only one of the reset options in jhi-programming-exercise-reset? If so, you should handle it there and reduce the duplication and unify similar components and behavior there

FelixTJDietrich replied:
Hmm, so you mean that I should just show the `Reset Programming Exercise` dialog with only the `recreate build plans` option available? All other options are unavailable for the Editor. I thought it would be nicer to just have the recreate build plans button right there instead of hidden in a dialog under reset.

JohannesStoehr replied:
I See advantages for both sides:
Easier code with less duplication and a more unified UI (e.g. an instructor is only an editor in another course) vs.
Simpler UI for editors.
Feel free to choose one of the options

JohannesStoehr commented on src/main/webapp/app/exercises/programming/manage/reset/programming-exercise-reset-button.component.ts:
```diff
@@ -0,0 +1,59 @@
+@Component({
+    selector: 'jhi-programming-exercise-reset',
+    template: `
+        <button
+            *ngIf="programmingExercise.isAtLeastInstructor"
+            [jhiFeatureToggle]="FeatureToggle.ProgrammingExercises"
+            class="btn btn-danger btn-sm"
+            style="display: flex; justify-content: center; align-items: center;"
+            (click)="openResetDialog($event)"
+            [ngClass]="{ 'd-inline-flex': !fullWidth }"
+            [ngStyle]="{
+                width: fullWidth ? '100%' : '',
+                'border-top-left-radius': noRoundedCornersTop ? '0' : '',
+                'border-top-right-radius': noRoundedCornersTop ? '0' : '',
+                'border-bottom-left-radius': noRoundedCornersBottom ? '0' : '',
+                'border-bottom-right-radius': noRoundedCornersBottom ? '0' : ''
+            }"
+        >
+            <div>
+                <fa-icon [icon]="faTimes"></fa-icon>
+                <span jhiTranslate="entity.action.reset">Reset</span>
+            </div>
+        </button>
+    `,
```
I'd use an extra file for this

JohannesStoehr commented on src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java:
```diff
@@ -625,6 +626,45 @@ public ResponseEntity<Void> recreateBuildPlans(@PathVariable Long exerciseId) {
+     */
+    @PutMapping(RESET)
+    @PreAuthorize("hasRole('INSTRUCTOR')")
+    public ResponseEntity<Void> reset(@PathVariable Long exerciseId, @RequestBody ProgrammingExerciseResetOptionsDTO programmingExerciseResetOptionsDTO) {
```
Since we have this new Endpoint, is it possible to delete the others that were only meant to reset certain things?

FelixTJDietrich replied:
It seems that I can safely remove exercise's `/cleanup` but exercise's `/reset` is used by quiz exercises. I will remove `/cleanup` since it seems anyways to be located in the wrong file.

pal03377 commented on src/main/webapp/app/exercises/programming/manage/reset/programming-exercise-reset-dialog.component.ts:
```diff
@@ -0,0 +1,128 @@
+     * @returns {boolean} true if the warning message should be displayed, false otherwise
+     */
+    showUndeletedArtifactsWarning() {
+        const options = [
+            this.programmingExerciseResetOptions.deleteBuildPlans,
+            this.programmingExerciseResetOptions.deleteRepositories,
+            this.programmingExerciseResetOptions.deleteParticipationsSubmissionsAndResults,
+        ];
+        return (!options[0] && (options[1] || options[2])) || (!options[1] && options[2]);
```
I find this way of writing it a bit confusing. Maybe you could add some helper variable called `opts` and then write the whole conditions for clarity?
```suggestion
        const opts = this.programmingExerciseResetOptions;
        return (!opts.deleteBuildPlans && (opts. deleteRepositories || opts. deleteParticipationsSubmissionsAndResults)) || (!opts. deleteRepositories && opts. deleteParticipationsSubmissionsAndResults);
```

FelixTJDietrich replied:
I now have it in the html as a simplified method. The warning now only appears when option 3 is set but 2 is unset.

pal03377 commented on src/main/webapp/app/exercises/programming/manage/reset/programming-exercise-reset-dialog.component.ts:
```diff
@@ -0,0 +1,128 @@
+        this.programmingExerciseService.reset(this.programmingExercise.id, this.programmingExerciseResetOptions).subscribe({
+            next: this.handleResetResponse,
+            error: () => {
+                this.resetInProgress = false;
```
Will this still show an error alert on failure? I think it should

pal03377 commented on src/main/webapp/app/exercises/programming/manage/reset/programming-exercise-reset-dialog.component.ts:
```diff
@@ -0,0 +1,128 @@
+     * if the programmingExercise.title and entered confirmation matches
+     * @returns {boolean} true if security checks are fulfilled, false otherwise
+     */
+    get areSecurityChecksFulfilled(): boolean {
```
It's just one
```suggestion
    get isSecurityCheckFulfilled(): boolean {
```

pal03377 commented on src/main/webapp/i18n/de/programmingExercise.json:
```diff
@@ -25,7 +25,20 @@
+                    "warning": "Archivierung der Übung vor Aktivierung dieser Option wird empfohlen."
+                },
+                "deleteParticipationsSubmissionsAndResults": "Lösche Teilnahmen, Abgaben und Ergebnisse der Studierenden (in Artemis).",
+                "artefactsWarning": "Das Auswählen der Option 3 ohne vorherige Optionen kann zu nicht gelöschten Artefakten auf {{ vcPlatform }} und {{ ciPlatform }} führen. Bitte stellen Sie sicher, dass dies die gewünschte Aktion ist, bevor Sie fortfahren.",
```
I think it would be better to call "Option 3" something different in case there will be more options before that in the future.

FelixTJDietrich replied:
Thanks, I changed it

pal03377 commented on src/test/java/de/tum/in/www1/artemis/programmingexercise/ProgrammingExerciseIntegrationBambooBitbucketJiraTest.java:
```diff
@@ -964,6 +964,56 @@ void testRecreateBuildPlansSuccess() throws Exception {
 
+    // Tests for reset endpoint
+
+    @Test
+    @WithMockUser(username = TEST_PREFIX + "student1", roles = "STUDENT")
+    void testResetForbiddenStudent() throws Exception {
+        programmingExerciseIntegrationTestService.testResetForbidden();
+    }
+
+    @Test
+    @WithMockUser(username = TEST_PREFIX + "tutor1", roles = "TA")
+    void testResetForbiddenTutor() throws Exception {
+        programmingExerciseIntegrationTestService.testResetForbidden();
+    }
+
+    @Test
+    @WithMockUser(username = TEST_PREFIX + "editor1", roles = "EDITOR")
+    void testResetForbiddenEditor() throws Exception {
+        programmingExerciseIntegrationTestService.testResetForbidden();
+    }
```
I think one check for the editor would suffice here.

pal03377 commented on src/test/java/de/tum/in/www1/artemis/programmingexercise/ProgrammingExerciseIntegrationJenkinsGitlabTest.java:
```diff
@@ -731,6 +731,54 @@ void recreateBuildPlansSuccess() throws Exception {
         programmingExerciseIntegrationTestService.testRecreateBuildPlansExerciseSuccess();
     }
 
+    @Test
+    @WithMockUser(username = TEST_PREFIX + "student1", roles = "STUDENT")
+    void testResetForbiddenStudent() throws Exception {
+        programmingExerciseIntegrationTestService.testResetForbidden();
+    }
+
+    @Test
+    @WithMockUser(username = TEST_PREFIX + "tutor1", roles = "TA")
+    void testResetForbiddenTutor() throws Exception {
+        programmingExerciseIntegrationTestService.testResetForbidden();
+    }
+
+    @Test
+    @WithMockUser(username = TEST_PREFIX + "editor1", roles = "EDITOR")
+    void testResetForbiddenEditor() throws Exception {
+        programmingExerciseIntegrationTestService.testResetForbidden();
+    }
```
```suggestion
    @Test
    @WithMockUser(username = TEST_PREFIX + "editor1", roles = "EDITOR")
    void testResetForbiddenEditor() throws Exception {
        programmingExerciseIntegrationTestService.testResetForbidden();
    }
```

pal03377 commented on src/test/javascript/spec/component/programming-exercise/programming-exercise-reset-dialog.component.spec.ts:
```diff
@@ -0,0 +1,234 @@
+        jest.restoreAllMocks();
+    });
+
+    it('test initialization', () => {
+        fixture.detectChanges();
+        expect(comp.programmingExercise.id).toBe(42);
+    });
```
```suggestion
```

I think this doesn't really test anything about the component itself (you do `comp.programmingExercise = programmingExercise` before

pal03377 commented on src/test/javascript/spec/component/programming-exercise/programming-exercise-reset-dialog.component.spec.ts:
```diff
@@ -0,0 +1,234 @@
+        expect(comp.programmingExercise.id).toBe(42);
+    });
+
+    it('Programming exercise service should find the correct programming exercise', () => {
```
```suggestion
    it('should find the correct programming exercise', () => {
```
I would be nicer if the `it` test description strings would always start with "should"

pal03377 commented on src/test/javascript/spec/component/programming-exercise/programming-exercise-reset-dialog.component.spec.ts:
```diff
@@ -0,0 +1,234 @@
+        expect(comp.programmingExercise).toEqual(programmingExercise);
+    });
+
+    it('clear() should close the modal dialog', () => {
```
same thing with "should"

pal03377 commented on src/test/javascript/spec/component/programming-exercise/programming-exercise-reset-dialog.component.spec.ts:
```diff
@@ -0,0 +1,234 @@
+            expect(comp.hasSelectedOptions).toBeFalse();
+        });
+
+        it('should return true when at least one option is set to true', () => {
+            const options = {
+                deleteBuildPlans: true,
+                deleteStudentRepositories: true,
+                deleteStudentParticipationsSubmissionsAndResults: true,
+                recreateBuildPlans: true,
+            };
+
+            for (const option in options) {
+                comp.programmingExerciseResetOptions = {
+                    deleteBuildPlans: false,
+                    deleteRepositories: false,
+                    deleteParticipationsSubmissionsAndResults: false,
+                    recreateBuildPlans: false,
+                };
+                comp.programmingExerciseResetOptions[option] = true;
+                expect(comp.hasSelectedOptions).toBeTrue();
+            }
+        });
```
Maybe you can create a parametrized test here instead?

FelixTJDietrich replied:
TIL, thanks!

pal03377 reviewed (changes requested):
I have a few requests for code changes. Also, I'm a bit worries about this being potentially incompatible with the upcoming local CI/VC systems.

FelixTJDietrich commented on src/main/java/de/tum/in/www1/artemis/service/connectors/bamboo/BambooBuildPlanLinkInfoContributor.java:
```diff
@@ -21,5 +21,8 @@ public class BambooBuildPlanLinkInfoContributor implements InfoContributor {
         final var buildPlanURLTemplate = bambooServerUrl + "/browse/{buildPlanId}";
         builder.withDetail(Constants.INFO_BUILD_PLAN_URL_DETAIL, buildPlanURLTemplate);
+
+        // Store name of the continuous integration system
+        builder.withDetail(Constants.CONTINUOUS_INTEGRATION_NAME, "Bamboo");
```
As far as I can see, no. If the info value is missing from that endpoint the UI just shows `{ ciPlatformName }`. We can just add an info contributor for the local CI/VC. But thanks for pointing that out, I just noticed that I should probably rename this file to `BambooInfoContributer.java`.

FelixTJDietrich commented on src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java:
```diff
@@ -625,6 +626,45 @@ public ResponseEntity<Void> recreateBuildPlans(@PathVariable Long exerciseId) {
+     * @param programmingExerciseResetOptionsDTO - Data Transfer Object specifying which operations to perform during the exercise reset.
+     * @return ResponseEntity<Void> - The ResponseEntity with status 200 (OK) if the reset was successful.
+     */
+    @PutMapping(RESET)
+    @PreAuthorize("hasRole('INSTRUCTOR')")
```
I will add it, I think maybe the whole route gets toggled, but not sure

FelixTJDietrich commented on src/main/webapp/app/exercises/programming/manage/reset/programming-exercise-reset-button.component.ts:
```diff
@@ -0,0 +1,59 @@
+            [ngClass]="{ 'd-inline-flex': !fullWidth }"
+            [ngStyle]="{
+                width: fullWidth ? '100%' : '',
+                'border-top-left-radius': noRoundedCornersTop ? '0' : '',
+                'border-top-right-radius': noRoundedCornersTop ? '0' : '',
+                'border-bottom-left-radius': noRoundedCornersBottom ? '0' : '',
+                'border-bottom-right-radius': noRoundedCornersBottom ? '0' : ''
```
Yeah, idk I will try, I was just frustrated that the bootstrap btn-group did not want to work for me with this button. I will try to make it nicer

FelixTJDietrich replied:
Fixed, not necessary anymore

FelixTJDietrich commented on src/main/webapp/app/exercises/programming/manage/programming-exercise.component.html:
```diff
@@ -243,20 +243,13 @@ <h4 class="mb-1 me-auto">
-                                >
-                                    <fa-icon [icon]="faTimes"></fa-icon>
-                                </button>
+                                    style="width: 100%"
```
Not necessary anymore. Removed.

sven0311 reviewed (dismissed):
Tested on TS2 with Gitlab and Jenkins. Worked as expected from the instructor perspective. Also the participation and score for the user was gone 👍

lennart-keller reviewed (dismissed):
Tested on ts2 within testing session as instructor. Works fine.

FelixTJDietrich commented on src/main/webapp/app/exercises/programming/manage/reset/programming-exercise-reset-dialog.component.html:
```diff
@@ -0,0 +1,118 @@
+            <fa-icon [icon]="faBan"></fa-icon>&nbsp;<span jhiTranslate="entity.action.cancel">Cancel</span>
+        </button>
+        <button
+            id="confirm"
```
Removed

FelixTJDietrich commented:
Thanks for all the suggestions, I've now implemented them all. @pal03377 thanks for being so thorough and critical with the code review 🙏

I also updated the testing steps to be hopefully easier to follow.

The only thing remaining is now the question with the editor role. I don't know what they do but I kept their recreate build plan button as before and did not hide it for them in a reset dialog.

pal03377 commented on src/main/webapp/app/exercises/programming/manage/reset/programming-exercise-reset-dialog.component.html:
```diff
@@ -0,0 +1,116 @@
+                    jhiTranslate="artemisApp.programmingExercise.reset.artefactsWarning"
+                    [translateValues]="{ vcPlatform: versionControlName, ciPlatform: continuousIntegrationName }"
+                >
+                    Selecting Options 3 without all previous options may lead to undeleted artifacts on {{ 'Bitbucket' }} and {{ 'Bamboo' }}. Please ensure this is your intended
+                    action before proceeding.
```
I think you forgot to update this text when updating the translation file

FelixTJDietrich replied:
Yeah also forgot to use the right names for the ci and vc, thanks for noticing!

pal03377 commented on src/main/webapp/app/exercises/shared/exercise/exercise.service.ts:
```diff
@@ -181,16 +181,6 @@ export class ExerciseService {
         );
```
As the GitHub action says, the `HttpParams` import above should be removed

pal03377 reviewed (approved):
The code looks really good, well done! 🎉  Also you were pretty quick with all the improvements

I just have two very small change requests left. Other than that, this lgtm 👍

FelixTJDietrich commented:
I implemented the requested changes and tested everything again on ts1 and ts2.

JohannesStoehr reviewed (dismissed):
Code changes look good to me

pal03377 commented on src/main/webapp/app/exercises/programming/manage/programming-exercise.component.html:
```diff
@@ -244,18 +244,15 @@ <h4 class="mb-1 me-auto">
+                                    jhiProgrammingExerciseResetButton
+                                    [programmingExercise]="programmingExercise"
                                     class="mb-1"
+                                    id="reset-exercise"
```
Is it never the case that `programming-exercise-detail.component` and this `programming-exercise.component` are used simultaneously somehow? Otherwise there could be a problem here with the HTML element ID clash (in the other component, there's also a button with `id="reset-exercise"`. If you are sure that's fine you can just resolve this.

FelixTJDietrich replied:
True, I removed it. Thanks!

pal03377 reviewed (approved):
Code changes still look good, however the changes since the last tests should definitely be re-tested thoroughly because quite a bit has changed

pal03377 reviewed (approved):
removal of id (since last review) also makes sense 👍

sven0311 reviewed (approved):
Tested on TS1. Still works as expected 👍

JohannesStoehr commented on src/main/webapp/i18n/de/programmingExercise.json:
```diff
@@ -25,7 +25,21 @@
+                },
+                "deleteParticipationsSubmissionsAndResults": "Lösche Teilnahmen, Abgaben und Ergebnisse der Studierenden (in Artemis).",
+                "artefactsWarning": "Das Löschen von Teilnahmen, Abgaben und Ergebnissen von Studierenden, ohne dabei die zugehörigen Repositories und Build-Pläne zu entfernen, kann zu nicht gelöschten Artefakten auf {{ vcPlatform }} und {{ ciPlatform }} führen. Bitte überprüfen Sie sorgfältig, ob dies die gewünschte Aktion ist, bevor Sie den Vorgang fortsetzen.",
+                "recreateBuildPlans": "Erstelle Übungs-Build-Pläne (BASE und SOLUTION) erneut und überschreibe frühere Änderungen.",
```
```suggestion
                "recreateBuildPlans": "Erstelle Build-Pläne der Aufgabe (BASE und SOLUTION) erneut und überschreibe frühere Änderungen.",
```

JohannesStoehr reviewed (approved):
Code changes look good to me. Left one suggestion for the German translation

TheZoker reviewed (dismissed):
Tested within testing session on TS2. LGTM 👍

rriyaldhi reviewed (approved):
Tested on TS2 and works as expected.

pellumbbaboci reviewed (approved):
Tested during testing session and its working as expected
